torch/csrc/jit/passes/quantization.cpp

```diff
@@ -697,18 +728,31 @@ void InsertQuantDeQuantHelper::removeObservers(
     }
     nodes_to_destroy_.at(g).clear();
   }
+
+  // If we have seen this graph before, we'll replay the observer
+  // slots removal using the slot index
+  if (removed_observer_slots_.count(g)) {
```

Comment:
This logic seems more complicated than it could be, or maybe some names are not descriptive enough.

Essentially, we want to remove observers attribute from the type and from objects of this type (there can be several of them). It looks to me that a simple `type_is_visited` or something like this set would reflect that logic better. I think `removed_observer_slots_` is playing this exact role, but I have to think hard and carefully read the code to verify that, which is not great. Separating cleanups for the type and for the objects might make it a bit easier to comprehend. The type cleanup will probably be still called from the object cleanup, but will be guarded by `if(type_cleanedup_.count(type))` check. The overall structure of the code would only slightly differ from what we have now.

Reply:
Do you mean that we don't cleanup the type in the beginning and only do that after all the slots are removed? I don't think that will work because we need to make sure objects and types are in sync for the slot index to make sense.
e.g.
Type A:
   attr1, attr2
object A:
   slot1, slot2
object B:
   slot1, slot2

if we want to remove both slot1 and slot2, and first we remove slot1 from object A, we find the index is 0 from querying type A, and we don't remove attr1 in type A. then object A and type A are not consistent anymore. when we want to remove slot2 from object A, and the query will return the index of 1 which does not exist anymore.